### PR TITLE
Use a methods struct for generated contracts

### PR DESCRIPTION
**Pull request by nlordell**
This PR adds a generated `{contract_name}Methods` struct that contains the implementation for all the contract methods instead of adding them to the root type. Additionally, `Deref<Target = Methods>` is implemented for the root contract type so calls to methods can still be done how they've always been done.

This is done so contract methods with names that conflict with other generated methods (such as `at`, `deployed`, etc.) can still be called this way by explicitly going through the `.methods()`.

### Test Plan

CI and make sure examples don't require code changed.

### CODE REVIEW

File: generate/src/contract/deployment.rs [hunk 1]
```diff
@@ -175,12 +171,7 @@ fn expand_deploy(cx: &Context) -> Result<TokenStream> {
             }
 
             fn at_address(web3: #ethcontract::DynWeb3, address: #ethcontract::Address, _: Self::Context) -> Self {
-                use #ethcontract::Instance;
-
-                let abi = Self::artifact().abi.clone();
-                Self {
-                    instance: Instance::at(web3, abi, address),
-                }
+                Self::at(&web3, address)
```
Comment (fleupold) [inline generate/src/contract/deployment.rs:174]: This method existed before, right? It's not added in this PR

File: generate/src/contract/methods.rs [hunk 2]
```diff
@@ -19,17 +21,48 @@ pub(crate) fn expand(cx: &Context) -> Result<TokenStream> {
...
+            /// would collide with a common method (like `at` or `deployed`).
+            pub fn methods(&self) -> &#methods {
+                &self.methods
+            }
+        }
+
+        /// Type containing all contract methods for generated contract type.
+        #[allow(non_camel_case_types)]
+        #[derive(Clone)]
+        pub struct #methods {
```
Comment (fleupold) [inline generate/src/contract/methods.rs]: Should this reuse `#methods_struct`?

**Review by fleupold — APPROVED** [review]
Looks great. Is this related to some concrete issue? Let's see when someone will use the first contract with such a naming collisions.

File: generate/src/contract/deployment.rs (hunk 1 above)
Comment (nlordell) [reply generate/src/contract/deployment.rs:174]: It changed because of how we achieve `Deref<Target = Methods>`.

**Comment by nlordell** [comment]
> Let's see when someone will use the first contract with such a naming collisions.

This is also related to events, contracts love to have events and mehods have the same name with different casing (for example ERC20 `Transfer` event). This causes issues when generating code and shoving it all into one type.

**Review by e00E — APPROVED** [review]

**Pull request merged**